accepted/future-releases/language-versioning/package-config-file-v2.md

```diff
@@ -0,0 +1,183 @@
+
+Some users may use other tools than Pub to control their package resolution. Those tools need to also support the new format.
+
+### Location
+
+The new file will be stored by Pub in the `.dart_tool` directory. It is a file used and shared by various Dart tools, so this location is sensible and consistent. It avoids clutter at the top-level of a package, and the `.dart_tool` directory is already in `.gitignore` files.
+
+Tools which needs the file will automatically look for it in `.dart_tool` subdirectories of the directories where they would currently look for `.package` files.
+
+### Name
+
+The new file's name should signal three things: 
+
+- It is a generated file, not intended for manual editing.
+- It is used by the Dart language tools to understand Dart source correctly.
+- It's JSON, so in case you do want to view or edit the file, the editor can treat it properly.
+
+Naming is still open. One option is "package_config.json".
+
+### Contents
+
+The file will contain the information that was planned for the `.packages` file:
+
+- A *default package* specified by its package name. This is the name of a package that non-package files will be considered part of for any purpose where belonging to a package is meaningful. It is intended to allow test and tool files in a Pub package to inherit any properties assigned to the corresponding Dart package (specifically language version for now, but potentially other features in the future).
+- A number of *packages*, each with:
+  - A *package name*.
+  - A *path* or *location URI* specifying where to find the files of the package.
+  - A *language version* (major/minor version number).
+
+On top of that, we also add extra metadata which allows us to version the format and attach information about when and how the file was created.
+
+The format is JSON text describing is a JSON object.
+
+That object has, at least, the following properties:
+
+- `apiVersion`: A single integer marking the version of this format, currently `2` (counting `.packages` as the first). Tools should refuse to work on files with a larger version number than the ones they are aware of. Incremental changes will not require increasing the version, so new versions are not expected any time soon, but code should still prepare for it.
+
+- `defaultPackage`: A string containing a package name.
+- `packages`: An array of packages, each with the following properties:
+  - `name`: A string containing a package name.
+  - either `path`: A string containing a relative or absolute path OS specific path using `/` as separator. If the path is relative, it is relative to the directory of the `package_config.json` file. 
+  - or `uri`: A URI reference. This is resolved relative to the (likely `file:`) URI referencing the `.package.json` file.
```

Comment:
I saw the discussion of _OS Path_ vs _URI_, and I see that `path` is OS specific, I think that's okay because this file is inherently OS specific, it can't be moved from one machine to another.

But I'm also okay to use `uri` instead of _OS path_, I suspect some tools will resort to regex parsing the URI, but that'll probably continue to work just as fine as it does now :see_no_evil: 

At-least using the `uri` property name will make it more obvious that it's a URI :)

Reply:
The path is not inherently OS specific, and in practice it needs to not be so.
The Dart SDK has a checked-in `.packages` file which works (and needs to wok) on any platform.